google/cloud/bigquery/job/query.py

```diff
@@ -989,19 +989,30 @@ def done(self, retry=DEFAULT_RETRY, timeout=None, reload=True):
 
         Returns:
             bool: True if the job is complete, False otherwise.
+
+        Raises:
+            google.api_core.exceptions,GoogleAPICallError:
+                If a non-retryable error ocurrs while reloading the job metadata,
+                or if too many retry attempts fail.
         """
         # Do not refresh if the state is already done, as the job will not
         # change once complete.
         is_done = self.state == _DONE_STATE
         if not reload or is_done:
             return is_done
 
-        self._reload_query_results(retry=retry, timeout=timeout)
-
         # If an explicit timeout is not given, fall back to the transport timeout
         # stored in _blocking_poll() in the process of polling for job completion.
         transport_timeout = timeout if timeout is not None else self._transport_timeout
 
+        try:
+            self._reload_query_results(retry=retry, timeout=transport_timeout)
+        except exceptions.GoogleAPIError:
+            # Reloading also updates error details on self, no need for an
+            # explicit self.set_exception() call.
+            self.reload(retry=retry, timeout=transport_timeout)
```

Comment:
Perhaps this should be changed to `exceptions.GoogleAPICallError` so that `RetryError` bubbles up immediately? Or could it still be beneficial to call `self.reload()` even if query results could not be fetched?

Reply:
> Or could it still be beneficial to call self.reload() even if query results could not be fetched?

Yes, this would be beneficial, as the job resource will contain extra information about why the job failed.

Reply:
OK, I'll leave it as is then, thanks.